cqc/pythonLib.py

```diff
@@ -184,7 +185,255 @@ def createXtraHeader(command, values):
+            xtra_msg = b""
+        else:
+            xtra_msg = xtra_hdr.pack()
+            header_length = CQCCmdHeader.HDR_LENGTH + xtra_hdr.HDR_LENGTH
+
+        # Construct Header
+        hdr = CQCHeader()
+        hdr.setVals(CQC_VERSION, CQC_TP_COMMAND, self._appID, header_length)
+        msg = hdr.pack()
+
+        # Construct Command
+        cmd_hdr = CQCCmdHeader()
+        cmd_hdr.setVals(qID, command, notify, block, action)
+        cmd_msg = cmd_hdr.pack()
+
+        return msg + cmd_msg + xtra_msg
+
+    def construct_release(self, qubits, notify=True, block=False, 
+                          action=False):
+        """Construct command to release qubits
+
+        - **Arguments**
+
+            :qubits:    a list of qubits to be released
+            :notify:    Do we wish to be notified when done.
+            :block:     Do we want the qubit to be blocked
+            :action:    Execute the releases recursively or sequencely
+        """
+
+        if isinstance(qubits, qubit):
+            qubits = [qubits]
+        assert isinstance(qubits, list)
+        n = len(qubits)
+
+        if n == 0:  # then we don't need to do anything
+            return
+
+        logging.debug("App {} tells CQC: Release {} qubits".format(self.name, n))
+        if action:
+            hdr_length = CQCCmdHeader.HDR_LENGTH + CQCSequenceHeader.HDR_LENGTH
+        else:
+            hdr_length = CQCCmdHeader.HDR_LENGTH
+
+        hdr = CQCHeader()
+        hdr.setVals(CQC_VERSION, CQC_TP_COMMAND, self._appID, hdr_length * n)
+        cqc_msg = hdr.pack()
+
+        release_messages = b""
+
+        for i in range(n):
+            q = qubits[i]
+            try:
+                q.check_active()
+            except QubitNotActiveError as e:
+                raise QubitNotActiveError(
+                    str(e) + ". Qubit {} is not active. None of the qubits are released".format(q._qID)
+                )
+            q._set_active(False)
+            cmd_hdr = CQCCmdHeader()
+            cmd_hdr.setVals(q._qID, CQC_CMD_RELEASE, int(notify), int(block), 
+                            int(action))
+            release_messages += cmd_hdr.pack()
+            if action:
+                seq_hdr = CQCSequenceHeader()
+                # After this one we are sending n-i-1 more releases
+                seq_hdr.setVals(hdr_length * (n - i - 1))
+                release_messages += seq_hdr.pack()
+
+        return cqc_msg + release_messages
+
+    def construct_simple(self, tp):
+        """Construct simple message.
+        
+        For example a HELLO message if tp=CQC_TP_HELLO.
+        """
+        hdr = CQCHeader()
+        hdr.setVals(CQC_VERSION, tp, self._appID, 0)
+        msg = hdr.pack()
+        return msg
+
+    def sendSimple(self, tp):
+        """Construct and commit simple message."""
+        
+        msg = self.construct_simple(tp)
+        self.commit(msg)
+
+    @abstractmethod
+    def new_qubitID(self):
```

Comment:
Both here and in the subclass it is not fully clear what this method is supposed to do. I think it would be good to have a bit more explanation in the doc-string.

Reply:
I've updated them. See this commit 0a292d23723c22ccc7a08a46f86a88796ab8b52b